File: backend/Core/OrchestrationGraph.cs
namespace DataNexus.Core;

public enum OrchestrationWorkflowKind
{
    Structured,
    Graph,
}

public sealed record OrchestrationGraph(
    IReadOnlyList<OrchestrationGraphNode> Nodes,
    IReadOnlyList<OrchestrationGraphEdge> Edges);

public sealed record OrchestrationGraphNode
{
    public string Id { get; init; } = string.Empty;
    public int DisplayOrder { get; init; }
    public string Title { get; init; } = string.Empty;
    public string Description { get; init; } = string.Empty;
    public int AgentId { get; init; }
    public string AgentName { get; init; } = string.Empty;
    public bool IsEdited { get; init; }
    public string? PromptOverride { get; init; }
    public Dictionary<string, string>? Parameters { get; init; }
    public double PositionX { get; init; }
    public double PositionY { get; init; }
}

public sealed record OrchestrationGraphEdge
{
    public string Id { get; init; } = string.Empty;
    public string SourceNodeId { get; init; } = string.Empty;
    public string TargetNodeId { get; init; } = string.Empty;
}

public static class OrchestrationGraphRules
{
    public static IReadOnlyList<OrchestrationStep> NormalizeSteps(IReadOnlyList<OrchestrationStep> steps)
    {
        if (steps.Count < 1)
            throw new InvalidOperationException("At least one step is required.");

        return steps
            .Select((step, index) => new OrchestrationStep
            {
                StepNumber = index + 1,
                Title = step.Title?.Trim() ?? string.Empty,
                Description = step.Description?.Trim() ?? string.Empty,
                AgentId = step.AgentId,
                AgentName = step.AgentName?.Trim() ?? string.Empty,
                IsEdited = step.IsEdited,
                PromptOverride = string.IsNullOrWhiteSpace(step.PromptOverride)
                    ? null
                    : step.PromptOverride.Trim(),
                Parameters = step.Parameters is { Count: > 0 }
                    ? new Dictionary<string, string>(step.Parameters, StringComparer.Ordinal)
                    : null,
            })
            .ToList();
    }

    public static OrchestrationGraph NormalizeGraph(OrchestrationGraph? graph)
    {
        if (graph is null)
            throw new InvalidOperationException("Graph data is required for graph orchestrations.");

        if (graph.Nodes.Count < 1)
            throw new InvalidOperationException("Graph orchestrations require at least one node.");

        var normalizedNodes = new List<OrchestrationGraphNode>(graph.Nodes.Count);
        var nodeIds = new HashSet<string>(StringComparer.Ordinal);

        foreach (var (node, index) in graph.Nodes.Select((item, idx) => (item, idx)))
        {
            var nodeId = node.Id?.Trim() ?? string.Empty;
            if (string.IsNullOrWhiteSpace(nodeId))
                throw new InvalidOperationException("Every graph node must have an id.");

            if (!nodeIds.Add(nodeId))
                throw new InvalidOperationException($"Duplicate graph node id '{nodeId}'.");

            normalizedNodes.Add(node with
            {
                Id = nodeId,
                DisplayOrder = node.DisplayOrder > 0 ? node.DisplayOrder : index + 1,
                Title = node.Title?.Trim() ?? string.Empty,
                Description = node.Description?.Trim() ?? string.Empty,
                AgentName = node.AgentName?.Trim() ?? string.Empty,
                PromptOverride = string.IsNullOrWhiteSpace(node.PromptOverride)
                    ? null
                    : node.PromptOverride.Trim(),
                Parameters = node.Parameters is { Count: > 0 }
                    ? new Dictionary<string, string>(node.Parameters, StringComparer.Ordinal)
                    : null,
            });
        }

        var normalizedEdges = new List<OrchestrationGraphEdge>(graph.Edges.Count);
        var edgeIds = new HashSet<string>(StringComparer.Ordinal);
        var edgePairs = new HashSet<string>(StringComparer.Ordinal);

        foreach (var (edge, index) in graph.Edges.Select((item, idx) => (item, idx)))
        {
            var edgeId = string.IsNullOrWhiteSpace(edge.Id) ? $"edge-{index + 1}" : edge.Id.Trim();
            if (!edgeIds.Add(edgeId))
                throw new InvalidOperationException($"Duplicate graph edge id '{edgeId}'.");

            var sourceNodeId = edge.SourceNodeId?.Trim() ?? string.Empty;
            var targetNodeId = edge.TargetNodeId?.Trim() ?? string.Empty;

            if (string.IsNullOrWhiteSpace(sourceNodeId) || string.IsNullOrWhiteSpace(targetNodeId))
                throw new InvalidOperationException($"Graph edge '{edgeId}' must have both source and target ids.");

            if (!nodeIds.Contains(sourceNodeId))
                throw new InvalidOperationException($"Graph edge '{edgeId}' references unknown source node '{sourceNodeId}'.");

            if (!nodeIds.Contains(targetNodeId))
                throw new InvalidOperationException($"Graph edge '{edgeId}' references unknown target node '{targetNodeId}'.");

            if (string.Equals(sourceNodeId, targetNodeId, StringComparison.Ordinal))
                throw new InvalidOperationException($"Graph edge '{edgeId}' cannot point a node to itself.");

            var edgePair = $"{sourceNodeId}->{targetNodeId}";
            if (!edgePairs.Add(edgePair))
                throw new InvalidOperationException($"Duplicate graph edge '{edgePair}'.");

            normalizedEdges.Add(edge with
            {
                Id = edgeId,
                SourceNodeId = sourceNodeId,
                TargetNodeId = targetNodeId,
            });
        }

        ValidateTopology(normalizedNodes, normalizedEdges);

        return new OrchestrationGraph(
            normalizedNodes.OrderBy(node => node.DisplayOrder).ToList(),
            normalizedEdges);
    }

    public static IReadOnlyList<OrchestrationStep> ToStructuredSteps(this OrchestrationGraph graph)
    {
        return graph.Nodes
            .OrderBy(node => node.DisplayOrder)
            .Select((node, index) => new OrchestrationStep
            {
                StepNumber = index + 1,
                Title = node.Title,
                Description = node.Description,
                AgentId = node.AgentId,
                AgentName = node.AgentName,
                IsEdited = node.IsEdited,
                PromptOverride = node.PromptOverride,
                Parameters = node.Parameters,
            })
            .ToList();
    }

    public static IReadOnlyList<string> GetRootNodeIds(OrchestrationGraph graph)
    {
        var incomingCounts = graph.Nodes.ToDictionary(node => node.Id, _ => 0, StringComparer.Ordinal);
        foreach (var edge in graph.Edges)
            incomingCounts[edge.TargetNodeId]++;

        return incomingCounts
            .Where(entry => entry.Value == 0)
            .Select(entry => entry.Key)
            .ToList();
    }

    public static IReadOnlyList<string> GetTerminalNodeIds(OrchestrationGraph graph)
    {
        var outgoingCounts = graph.Nodes.ToDictionary(node => node.Id, _ => 0, StringComparer.Ordinal);
        foreach (var edge in graph.Edges)
            outgoingCounts[edge.SourceNodeId]++;

        return outgoingCounts
            .Where(entry => entry.Value == 0)
            .Select(entry => entry.Key)
            .ToList();
    }

    private static void ValidateTopology(
        IReadOnlyList<OrchestrationGraphNode> nodes,
        IReadOnlyList<OrchestrationGraphEdge> edges)
    {
        var roots = GetRootNodeIds(new OrchestrationGraph(nodes, edges));
        if (roots.Count != 1)
            throw new InvalidOperationException("Graph orchestrations currently require exactly one start node.");

        var terminals = GetTerminalNodeIds(new OrchestrationGraph(nodes, edges));
        if (terminals.Count != 1)
            throw new InvalidOperationException("Graph orchestrations currently require exactly one terminal node.");

        var adjacency = nodes.ToDictionary(
            node => node.Id,
            _ => new List<string>(),
            StringComparer.Ordinal);
        var indegree = nodes.ToDictionary(
            node => node.Id,
            _ => 0,
            StringComparer.Ordinal);

        foreach (var edge in edges)
        {
            adjacency[edge.SourceNodeId].Add(edge.TargetNodeId);
            indegree[edge.TargetNodeId]++;
        }

        var reachable = new HashSet<string>(StringComparer.Ordinal);
        var traversalQueue = new Queue<string>();
        traversalQueue.Enqueue(roots[0]);
        reachable.Add(roots[0]);

        while (traversalQueue.Count > 0)
        {
            var nodeId = traversalQueue.Dequeue();
            foreach (var targetNodeId in adjacency[nodeId])
            {
                if (reachable.Add(targetNodeId))
                    traversalQueue.Enqueue(targetNodeId);
            }
        }

        if (reachable.Count != nodes.Count)
            throw new InvalidOperationException("Every graph node must be reachable from the start node.");

        var queue = new Queue<string>(indegree.Where(entry => entry.Value == 0).Select(entry => entry.Key));
        var visitedCount = 0;

        while (queue.Count > 0)
        {
            var nodeId = queue.Dequeue();
            visitedCount++;

            foreach (var targetNodeId in adjacency[nodeId])
            {
                indegree[targetNodeId]--;
                if (indegree[targetNodeId] == 0)
                    queue.Enqueue(targetNodeId);
            }
        }

        if (visitedCount != nodes.Count)
            throw new InvalidOperationException("Graph orchestrations must be acyclic in v1.");
    }
}

File: frontend/src/utils/orchestrationGraph.ts
import type {
  Agent,
  OrchestrationGraph,
  OrchestrationGraphEdge,
  OrchestrationGraphNode,
  OrchestrationStep,
} from "@/types/api";

const makeId = (prefix: string) => {
  if (typeof crypto !== "undefined" && typeof crypto.randomUUID === "function") {
    return `${prefix}-${crypto.randomUUID()}`;
  }

  return `${prefix}-${Date.now()}-${Math.random().toString(36).slice(2, 10)}`;
};

export function cloneGraph(graph: OrchestrationGraph): OrchestrationGraph {
  return {
    nodes: graph.nodes.map((node) => ({
      ...node,
      parameters: node.parameters ? { ...node.parameters } : null,
    })),
    edges: graph.edges.map((edge) => ({ ...edge })),
  };
}

export function graphToSteps(graph: OrchestrationGraph): OrchestrationStep[] {
  return [...graph.nodes]
    .sort((left, right) => left.displayOrder - right.displayOrder)
    .map((node, index) => ({
      stepNumber: index + 1,
      title: node.title,
      description: node.description,
      agentId: node.agentId,
      agentName: node.agentName,
      isEdited: node.isEdited,
      promptOverride: node.promptOverride,
      parameters: node.parameters,
    }));
}

export function buildGraphFromSteps(steps: OrchestrationStep[]): OrchestrationGraph {
  const nodes = steps.map<OrchestrationGraphNode>((step, index) => ({
    id: makeId("node"),
    displayOrder: index + 1,
    title: step.title,
    description: step.description,
    agentId: step.agentId,
    agentName: step.agentName,
    isEdited: step.isEdited,
    promptOverride: step.promptOverride,
    parameters: step.parameters,
    positionX: 120 + index * 220,
    positionY: 120,
  }));

  const edges = nodes.slice(0, -1).map<OrchestrationGraphEdge>((node, index) => ({
    id: makeId("edge"),
    sourceNodeId: node.id,
    targetNodeId: nodes[index + 1]!.id,
  }));

  return { nodes, edges };
}

export function createGraphNode(agent?: Agent, order = 1): OrchestrationGraphNode {
  return {
    id: makeId("node"),
    displayOrder: order,
    title: agent?.name ?? "New Node",
    description: "",
    agentId: agent?.id ?? 0,
    agentName: agent?.name ?? "",
    isEdited: true,
    promptOverride: null,
    parameters: null,
    positionX: 120 + (order - 1) * 180,
    positionY: 120,
  };
}

export function summarizeGraph(graph: OrchestrationGraph): Array<{ id: string; label: string }> {
  const nodeLookup = new Map(graph.nodes.map((node) => [node.id, node]));
  return graph.edges.map((edge) => ({
    id: edge.id,
    label: `${nodeLookup.get(edge.sourceNodeId)?.title ?? edge.sourceNodeId} -> ${nodeLookup.get(edge.targetNodeId)?.title ?? edge.targetNodeId}`,
  }));
}

export function validateGraph(graph: OrchestrationGraph): string[] {
  if (graph.nodes.length < 1) {
    return ["Add at least one node."];
  }

  const issues: string[] = [];
  const nodeIds = new Set<string>();
  const indegree = new Map<string, number>();
  const outgoing = new Map<string, string[]>();
  const edgePairs = new Set<string>();

  for (const node of graph.nodes) {
    if (!node.id.trim()) {
      issues.push("Every node needs an id.");
      continue;
    }
    if (nodeIds.has(node.id)) {
      issues.push(`Duplicate node id: ${node.id}`);
    }
    nodeIds.add(node.id);
    indegree.set(node.id, 0);
    outgoing.set(node.id, []);
  }

  for (const edge of graph.edges) {
    if (!nodeIds.has(edge.sourceNodeId) || !nodeIds.has(edge.targetNodeId)) {
      issues.push(`Edge ${edge.id} references a missing node.`);
      continue;
    }
    if (edge.sourceNodeId === edge.targetNodeId) {
      issues.push(`Edge ${edge.id} cannot connect a node to itself.`);
      continue;
    }
    const pair = `${edge.sourceNodeId}->${edge.targetNodeId}`;
    if (edgePairs.has(pair)) {
      issues.push(`Duplicate edge ${pair}.`);
      continue;
    }
    edgePairs.add(pair);
    indegree.set(edge.targetNodeId, (indegree.get(edge.targetNodeId) ?? 0) + 1);
    outgoing.get(edge.sourceNodeId)!.push(edge.targetNodeId);
  }

  const roots = graph.nodes.filter((node) => (indegree.get(node.id) ?? 0) === 0);
  if (roots.length !== 1) {
    issues.push("Graph workflows currently require exactly one start node.");
  }

  const terminals = graph.nodes.filter((node) => (outgoing.get(node.id) ?? []).length === 0);
  if (terminals.length !== 1) {
    issues.push("Graph workflows currently require exactly one terminal node.");
  }

  const queue = [...graph.nodes.filter((node) => (indegree.get(node.id) ?? 0) === 0).map((node) => node.id)];
  const indegreeCopy = new Map(indegree);
  let visited = 0;

  while (queue.length > 0) {
    const nodeId = queue.shift()!;
    visited += 1;

    for (const targetId of outgoing.get(nodeId) ?? []) {
      const nextCount = (indegreeCopy.get(targetId) ?? 0) - 1;
      indegreeCopy.set(targetId, nextCount);
      if (nextCount === 0) {
        queue.push(targetId);
      }
    }
  }

  if (visited !== graph.nodes.length) {
    issues.push("Graph workflows must be acyclic in v1.");
  }

  if (roots.length === 1) {
    const reachable = new Set<string>([roots[0]!.id]);
    const traversal = [roots[0]!.id];

    while (traversal.length > 0) {
      const nodeId = traversal.shift()!;
      for (const targetId of outgoing.get(nodeId) ?? []) {
        if (!reachable.has(targetId)) {
          reachable.add(targetId);
          traversal.push(targetId);
        }
      }
    }

    if (reachable.size !== graph.nodes.length) {
      issues.push("Every node must be reachable from the start node.");
    }
  }

  return [...new Set(issues)];
}